static/js/input.js

```diff
@@ -446,42 +445,32 @@ function guessColumnType(arr, threshold = 600) {
 
 }
 
-
 /**
  * Calculates the entropy of an array
  * @function calcEntropy
  * @param {Array} arr - array of values
  * @returns {Number} entropy - entropy of array
  */
-function calcEntropy(arr, match) {
-  const unique = [...new Set(arr)];
-  const base = unique.length;
-  let probs = [];
-
-  // calculates probability of occurence of search value
-  function calcProb(match) {
-    const n = arr.length;
-
-    let count = 0;
-    for (let i = 0; i < n; i++) {
-      if (arr[i] === match) {
-        count++;
-      }
+function calcEntropy(arr) {
```

Comment:
Here you use one `for` loop to calculate probabilities of all unique values. It is a great improvement! There are still a few more things you can learn regarding algorithm and optimization.

You defined `unique` as an array, and you used the `indexOf` method to check whether a value already exists. It isn't the most efficient way, because searching an array for a value is linear (_O_(_n_)). That is, you will need to loop the array again in order to find a value.

What I used in the `listCat` function is an object. The underlying implementation of a JavaScript object is a [hash table](https://en.wikipedia.org/wiki/Hash_table). Hash table is an interesting data structure, which allows constant search time (_O_(1)). That is, no matter how big the object is, you always gets the desired element in a flash! When the size is big, the advantage of object over array is significant.

Another trick for optimization: you can do `const n = arr.length` before the `for` loop, then do `i < n` in the loop. This will avoid repeated accession of the `length` attribute of `arr` (which cost some resource). In modern web browsers, this trick may not improve performance much, but it is good to know and adopt as a habit.

A trivial point: `indexOf` returns -1 if not found. For sake of specificity, you should use `=== -1` instead of `< 0` when you check the presence of the value.